### ISSUE TITLE
Configuration of KL divergence weight, gradient clipping and importance ratio clipping

### ISSUE THREAD

**Issue by faustomilletari** [PrimeIntellect-ai/prime-rl#877]
Hello,

I am observing a behavior during training where my rewards are increasing for a number of steps but inexplicably drop after a certain point while the model starts producing gibberish. 

I believe this might be due to the fact that the model is straying too far away from its original policy. 

I'd like to increase the weight of the KL divergence term in GRPO or operate gradient clipping or importance ratio clipping as ways to avoid this issue.

I was expecting to be able to pass these parameters in training.toml but I am unsure this is even possible.

Any feedback is appreciated.

**Comment by mikasenghaas** [comment]
we have dropped the kl loss term a while back as most recent literature has also dropped it. you should be able to configure gradient clipping via `--optim.max-norm` on the trainer and importance ratio config as `--loss.clip_ratio`. hope this helps!

**Comment by faresobeid** [comment]
Could you send your wandb logs?

**Comment by faustomilletari** [comment]
Hello,

thanks for the pointers in regard to the additional parameters I can pass.

I'll try to share my wandb logs, but another things I have noticed is that the issue I am experiencing only happens when I use tools in my environment. 

I am now doing a new run without tools, and the issue disappeared. I am going to double check my tool implementation and its effect on training. From the rollouts I can see that the tools are used appropriately and return correct results.

**Comment by faresobeid** [comment]
Thanks for the reply!
Are the tools always being used correctly? Could check out maybe https://arxiv.org/abs/2509.02479 which finds some issues like u mention with tool use

**Comment by faustomilletari** [comment]
from the paper I read:

>... masking the loss on these feedback tokens is a known practice (Jin et al., 2025, Mai et al., 2025)...

I assume this masking is controlled by: "mask_env_responses" which is an orchestrator parameter

**Comment by faresobeid** [comment]
Yes that's correct, although we don't have the masking of void turns done here. Will add it as an option but could be worth trying and seeing if it fixes the issue
